connectors/src/connectors/google_drive/temporal/workflows.ts

```diff
@@ -53,26 +54,36 @@ export async function googleDriveFullSync(
     signaled = false;
     let totalCount = 0;
     let nextPageToken: string | undefined = undefined;
-    do {
-      if (signaled) {
-        console.log(
-          "Folders selection changed, should start the sync all over again."
-        );
-        break;
+    let foldersToBrowse: string[] = await getFoldersToSync(connectorId);
+    while (foldersToBrowse.length > 0) {
+      const folderId = foldersToBrowse.pop();
```

Comment:
Given the fact that there is no absolute guarantee that we have a DAG that means we may end up in an infinite loop.

I think we want to store folders that we traverse and skip the ones already synced in that run?

Reply:
Yes I actually thought about it but I was a bit scared by Temporal having to serialize the big list of visited folders already visited each time we visit one (to save the workflow state).
I did run a little workflow test and observed no added latency even after a lot of strings being saved in memory.
```
export async function testLargeLocalState(connectorId: ModelId) {
  const queue: string[] = [];
  for (let i = 0; i < 100000; i++) {
    const start = new Date();
    for (let j = 0; j < 10000; j++) queue.push(`item-${i}-${j}`);
    await syncSucceeded(connectorId);
    const finish = new Date();
    console.log(
      `syncSucceeded took ${finish.getTime() - start.getTime()}ms - size: ${
        queue.length
      }`
    );
  }
}

```

I am pushing another commit right now.

Reply:
I was more thinking about storing them in DB. Will be useful to show permissions as well ?